### PR TITLE
Release 1.0.1

### PR DESCRIPTION
**Pull request by JanHolger**
Added:
- New self-implemented JsonParser and JsonDumper implementation
- New self-implemented Mapper
- Support for ordering fields using the order attribute in `@MapperOptions`
- Support for additional attributes by adding `@Additional` to a field of type AbstractObject
- Support for deserializing types with generic type arguments (Set, List, Map)
- Support for accessing the context (field, annotations, ...) in custom type adapters
- Stream API collectors for AbstractArray and AbstractObject that can be accessed using the static methods `AbstractArray.collect()` and `AbstractObject.collect()`
- Bson Support (bson library is marked optional and needs to be added to the classpath manually)

Fixed:
- QueryString keys weren't encoded/decoded properly
- Json String length limit of 1500

Deprecated:
- AbstractMapper (see the Mapper class for the new api)

Removed:
- Gson dependency
- Gson util classes (GsonAbstractDataAdapter, GsonEnum, GsonEnumAdapter, GsonListDeserializer, GsonMapDeserializer)

Required Changes:
- The old `AbstractMapper` got deprecated and should be replaced by the new `Mapper`
- Because gson isn't used for mapping anymore, you need to make sure to replace all gson annotations (`@SerializedName`, `@Expose`, ...) to `@MapperOptions`

### CODE REVIEW

File: src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java [hunk 1]
```diff
@@ -97,6 +58,14 @@ public Object fromAbstract(MapperContext context, AbstractElement element, Class
                     throw new MapperWrongTypeException(context.getField().getName(), "string", Helpers.typeName(element));
                 return element.string();
             }
+            if(type.equals(char.class) || type.equals(Character.class)) {
+                if(!element.isBoolean())
```
Comment (x7airworker) [inline src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java]: Can you explain why this is checking for a boolean?

File: src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java [hunk 2]
```diff
@@ -161,6 +153,24 @@ public Object fromAbstract(MapperContext context, AbstractElement element, Class
             }
         }
 
+        public Class<?>[] getSupportedTypes() {
```
Comment (x7airworker) [inline src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java:156]: Maybe Stack and Vector should also be added?

File: src/main/java/org/javawebstack/abstractdata/bson/BsonTypeAdapter.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    public BsonTypeAdapter(BsonConverter converter) {
+        this.converter = converter;
+    }
+
+    public AbstractElement toAbstract(MapperContext context, Object value) throws MapperException {
+        if(value instanceof byte[]) {
+            ByteBuffer buffer = ByteBuffer.wrap((byte[]) value);
+            buffer.order(ByteOrder.LITTLE_ENDIAN);
```
Comment (x7airworker) [inline src/main/java/org/javawebstack/abstractdata/bson/BsonTypeAdapter.java:33]: Is this defined by the spec or should it be changable to `BIG_ENDIAN` also?

**Review by x7airworker — CHANGES_REQUESTED** [review]

File: src/main/java/org/javawebstack/abstractdata/bson/BsonTypeAdapter.java (hunk 3 above)
Comment (JanHolger) [reply src/main/java/org/javawebstack/abstractdata/bson/BsonTypeAdapter.java:33]: This is defined by spec (see: https://bsonspec.org/spec.html) and i derived this logic from the bson internals. Unfortunately the bson library doesn't seem to provide any documented and easy to use api to actually convert the document to bson so we need to use internal api's here.

File: src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java (hunk 1 above)
Comment (JanHolger) [reply src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java]: fixed

File: src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java (hunk 2 above)
Comment (JanHolger) [reply src/main/java/org/javawebstack/abstractdata/mapper/DefaultMappers.java:156]: added

**Review by x7airworker — APPROVED** [review]

**Pull request merged**